File: contracts/wasm/tokenregistry/go/tokenregistry/tokenregistry.go
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

package tokenregistry

import (
	"github.com/iotaledger/wasp/packages/wasmvm/wasmlib/go/wasmlib"
)

func funcMintSupply(ctx wasmlib.ScFuncContext, f *MintSupplyContext) {
	minted := ctx.Minted()
	mintedTokens := minted.TokenIDs()
	ctx.Require(len(mintedTokens) == 1, "need single minted color")
	mintedColor := mintedTokens[0]
	currentToken := f.State.Registry().GetToken(*mintedColor)
	if currentToken.Exists() {
		// should never happen, because transaction id is unique
		ctx.Panic("TokenRegistry: registry for color already exists")
	}
	token := &Token{
		Supply:      minted.Balance(mintedColor).Uint64(),
		MintedBy:    ctx.Caller(),
		Owner:       ctx.Caller(),
		Created:     ctx.Timestamp(),
		Updated:     ctx.Timestamp(),
		Description: f.Params.Description().Value(),
		UserDefined: f.Params.UserDefined().Value(),
	}
	if token.Description == "" {
		token.Description += "no dscr"
	}
	currentToken.SetValue(token)
	colorList := f.State.TokenList()
	colorList.AppendTokenID().SetValue(*mintedColor)
}

func funcTransferOwnership(ctx wasmlib.ScFuncContext, f *TransferOwnershipContext) {
	// TODO
}

func funcUpdateMetadata(ctx wasmlib.ScFuncContext, f *UpdateMetadataContext) {
	// TODO
}

func viewGetInfo(ctx wasmlib.ScViewContext, f *GetInfoContext) {
	// TODO
}

File: contracts/wasm/tokenregistry/ts/tokenregistry/tokenregistry.ts
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import * as wasmlib from "wasmlib"
import * as sc from "./index";

export function funcMintSupply(ctx: wasmlib.ScFuncContext, f: sc.MintSupplyContext): void {
    let minted = ctx.minted();
    let mintedTokens = minted.tokenIDs();
    ctx.require(mintedTokens.length == 1, "need single minted color");
    let mintedColor = mintedTokens[0];
    let currentToken = f.state.registry().getToken(mintedColor);
    if (currentToken.exists()) {
        // should never happen, because transaction id is unique
        ctx.panic("TokenRegistry: registry for color already exists");
    }
    let token = new sc.Token();
    token.supply = minted.balance(mintedColor).uint64();
    token.mintedBy = ctx.caller();
    token.owner = ctx.caller();
    token.created = ctx.timestamp();
    token.updated = ctx.timestamp();
    token.description = f.params.description().value();
    token.userDefined = f.params.userDefined().value();
    if (token.description == "") {
        token.description = "no dscr";
    }
    currentToken.setValue(token);
    let colorList = f.state.tokenList();
    colorList.appendTokenID().setValue(mintedColor);
}

export function funcTransferOwnership(ctx: wasmlib.ScFuncContext, f: sc.TransferOwnershipContext): void {
    // TODO
}

export function funcUpdateMetadata(ctx: wasmlib.ScFuncContext, f: sc.UpdateMetadataContext): void {
    // TODO
}

export function viewGetInfo(ctx: wasmlib.ScViewContext, f: sc.GetInfoContext): void {
    // TODO
}

export function funcInit(ctx: wasmlib.ScFuncContext, f: sc.InitContext): void {
	if (f.params.owner().exists()) {
		f.state.owner().setValue(f.params.owner().value());
		return;
	}
	f.state.owner().setValue(ctx.requestSender());
}
